### PR TITLE
Add fuyu model

### PR DESCRIPTION
**Pull request by molbap**
# What does this PR do?

This PR adds the model Fuyu from Adept, a LMM processing image and text inputs.

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
LGTM if including the changes on the other PR

**Review by ArthurZucker — APPROVED** [review]
LGTM waiting for the CI to merge

File: src/transformers/models/fuyu/modeling_fuyu.py
Comment (NielsRogge) [inline src/transformers/models/fuyu/modeling_fuyu.py:158]: FYI this term here (`self.language_model`) needs to correspond to the `base_model_prefix` defined in `FuyuPreTrainedModel`. This ensures that when people do:

```
from transformers import FuyuModel

model = FuyuModel.from_pretrained("adept/fuyu-8b")
```

, that weights can be loaded properly for the base model (which does not include the language modeling head on top).

Comment (NielsRogge) [reply src/transformers/models/fuyu/modeling_fuyu.py:158]: Currently I see that the base_model_prefix is defined to be "fuyu", so that would need an update

File: src/transformers/models/fuyu/modeling_fuyu.py
Comment (NielsRogge) [inline src/transformers/models/fuyu/modeling_fuyu.py:218]: FYI for Pix2Struct we used the term [flattened_patches](https://github.com/huggingface/transformers/blob/5a73316bed96e9e022f9679dc8d1a7487f4721e1/src/transformers/models/pix2struct/modeling_pix2struct.py#L1644C16-L1644C16) => would be great to have consistent naming

File: src/transformers/models/fuyu/image_processing_fuyu.py
Comment (NielsRogge) [inline src/transformers/models/fuyu/image_processing_fuyu.py:25]: wait, this PR gets an approval with the current image processor? We're lucky Amy isn't seeing this :D

On a more serious note, the image processor needs a big rework to make it compliant with all our existing image processors (e.g. the size argument needs to be a dictionary by default, e.g. size = {"height": 1080, "width": 1920}, arguments like `do_resize` and `do_pad` need to be added, a separate preprocess function needs to be implemented, etc.

See also `Pix2StructImageProcessor` as a reference example

Comment (ArthurZucker) [reply src/transformers/models/fuyu/image_processing_fuyu.py:25]: For sure 🙌 but it’s a time sensitive release gotta merge now and refactor after

Comment (ArthurZucker) [reply src/transformers/models/fuyu/image_processing_fuyu.py:25]: The processor needs to be refactored as well and the config for default arguments as well + no TestMixin

Comment (ArthurZucker) [reply src/transformers/models/fuyu/image_processing_fuyu.py:25]: Might be a few but there won’t be a release before this is fixed so should be alright I hope.

File: src/transformers/models/fuyu/modeling_fuyu.py
Comment (ArthurZucker) [reply src/transformers/models/fuyu/modeling_fuyu.py:158]: there is no FuyuModel, it's PerisommonForCausalLM that would need to be loaded here but it won't have access to this anyway

File: tests/models/fuyu/test_image_processing_fuyu.py
Comment (ArthurZucker) [inline tests/models/fuyu/test_image_processing_fuyu.py:13]:
```suggestion
if is_torch_available() and is_vision_available():
```

File: tests/models/fuyu/test_processing_fuyu.py
Comment (ArthurZucker) [inline tests/models/fuyu/test_processing_fuyu.py:21]: reauire vision as well

File: tests/models/fuyu/test_processing_fuyu.py
Comment (ArthurZucker) [inline tests/models/fuyu/test_processing_fuyu.py:93]: reauire vision as well

File: src/transformers/models/fuyu/image_processing_fuyu.py
Comment (molbap) [reply src/transformers/models/fuyu/image_processing_fuyu.py:25]: Yeah the ImageProcessor has to be reworked, will be easier to do once we have the model up and running as well. I'll continue working on it and make it aligned with the rest :)

**Comment by phillip-kravtsov** [comment]
Hello! Thanks for supporting this model. Unfortunately, loading the model as per the hub directions seems to fail on `main`:
`model = transformers.FuyuForCausalLM.from_pretrained("adept/fuyu-8b", device_map="cuda:0", torch_dtype=torch.float16)`
The code warns that every tensor is unused, so it initializes every single tensor. I can confirm it works on [this commit](https://github.com/huggingface/transformers/pull/26911/commits/c36b15b88d8096551d9bdef710965dc648b7d893), though I haven't found the specific commit that breaks.

Additionally, the processor does not seem to support batch_size > 1 inputs correctly, the code  appears to construct the tensors correctly, but then slices to take their first element only (except the patch indices) [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/fuyu/processing_fuyu.py#L545-L546).

File: src/transformers/models/fuyu/image_processing_fuyu.py
Comment (NielsRogge) [reply src/transformers/models/fuyu/image_processing_fuyu.py:25]: Ok :)

**Comment by NightMachinery** [comment]
How much VRAM does Fuyu-8B need to run? 

Is it possible to have quantized versions running in C like the models TheBloke provides?

**Comment by ArthurZucker** [comment]
1. You need around 20GB for the bfloat16
2. You can already use the quantisation config with the integration in transformers see [the doc here](https://huggingface.co/docs/transformers/main_classes/quantization). 
3. This kind of question would be best asked directly on the model [on the hub](https://huggingface.co/adept/fuyu-8b/discussions)!

**Pull request merged**
